content/article/enhancing-midi-hardware-with-perl.md

```diff
@@ -0,0 +1,510 @@
+                    sub( $ts, $msg, $event, $data ) {
+                        $midi_ch->send( $event );
+                    }
+                );
+
+                sleep;
+            }
+        );
+        $loop->add( $midi_rtn );
+        $loop->await( $self->_process_midi_events );
+    }
+
+    ADJUST {
+        $self->_init_out;
+    }
+```
+
+The `_init_out` method takes care of some shortcomings in Windows MIDI, which
+does not support the creation of virtual ports. On this platform messages will
+be routed via [loopMIDI](https://www.tobias-erichsen.de/software/loopmidi.html).
+On other platforms the virtual MIDI port "RtMidi Output Client:Mister Fancy Pants" is created.
+The `ADJUST` block assures this is done during construction of the `MidiFilter`
+instance.
+
+The `go` method creates a routine which instantiates a RtMidi instance, and
+connects to the hardware MIDI device specified on the command line. A callback is
+created to send incoming events over the communications channel, then we
+simply sleep and allow RtMidi's event loop to take over the routine.
+
+The final step is to await `_process_midi_events`, which should process events
+from the hardware until the program is terminated.
+
+### Writing Callbacks
+
+Callbacks are responsible for managing the stash, and sending filtered messages
+to the output device. A callback receives the MidiFilter instance and the
+incoming event.
+
+In order to implement the pedal tone feature described earlier, we need to take
+incoming "note on" events and transform them into three "note on" events, then
+send these to the output MIDI device. A similar filter is needed for "note off"
+— all three notes must be stopped after being played:
+
+```perl
+use constant PEDAL => 55; # G below middle C
+
+sub pedal_notes( $note ) {
+    ( PEDAL, $note, $note + 7 );
+}
+
+sub pedal_tone( $mf, $event ) {
+    my ( $ev, $channel, $note, $vel ) = $event->@*;
+    $mf->send( [ $ev, $channel, $_, $vel ] ) for pedal_notes( $note );
+    true;
+}
+
+my $mf = MidiFilter->new;
+
+$mf->add_filter( note_on  => \&pedal_tone );
+$mf->add_filter( note_off => \&pedal_tone );
+
+$mf->go;
+```
+
+We start by setting a constant containing a MIDI note value for the pedal tone.
+The sub `pedal_notes` returns this pedal tone, the played note, and its fifth.
+The callback function `pedal_tone` sends a MIDI message to output for each of
+the notes returned by `pedal_notes`.
+Note the callback yields `true` in order to prevent falling through to
+the default action.
+The callback function is applied to both the "note on" and "note off" events.
+We finish by calling the `go` method of our `MidiFilter` instance in order to
+await and process incoming messages from the keyboard.
+
+The last step is to run the script:
+
+```
+$ ./midi-filter.pl ^oxy
+```
+
+Rather than specify a fully qualified device name, we can pass in a regex which
+should match any device whose name starts with "oxy" - there is only one match
```

Comment:
I would remind the reader that "oxy" relates to your "M-Audio Oxygen" device, here.